ha/util/message_bus.py

```diff
@@ -142,6 +143,112 @@ def stop(self):
+
+class MessageBusTransactionalConsumer:
+    CONSUMER = "CONSUMER"
+
+    def __init__(self, consumer_name: str, group_id: str, message_type: str, callback: Callable):
+        """
+        Initalize consumer.
+        Args:
+            consumer_name (str): should be unique at consumer level
+            group_id (str): unique at group level
+            message_type (list): Message Type.
+            callback (Callable): callback function to process message.
+        """
+        self.consumer_name = consumer_name
+        self.group_id = group_id
+        self.message_type = message_type
+        self.callback = callback
+        self.comm_type = MessageBusTransactionalConsumer.CONSUMER
+        self.stop_thread = False
+
+    def run(self):
+        """
+        Overloaded of Thread.
+        Note: Please properly handle failure cases to avoid stuck in loop
+
+        1. Caller received and processed message		                            SUCCESS
+        2. Caller received but not able to process need retry			            FAILED
+        3. Caller received but message is irrelevant to caller					    SUCCESS
+        4. Caller received	message but failed to process and not want to retry		FAILED_STOP
+        5. Caller received message and want to sop listen to message				SUCCESS_STOP
+        6. If nothing is passed it will be case 2						            FAILED
+        7. Exception will be swallowed and ack.
+        8. Closing main thread will close this thread as it is running as deamon.
+
+        As self.consumer.receive(timeout) is block call t1.join() will not stop thread.
+        Stop thread by completing work as per above cases.
+        """
+        retry = False
+        while not self.stop_thread:
+            try:
+                if not retry:
+                    message = self.consumer.recv(TOPIC=[self.message_type])
+                try:
+                    status = self.callback(message)
+                except Exception as e:
+                    Log.error(f"Caught exception from caller: {e}. retry again ...")
+                    retry = True
+                    continue
+                if status == CONSUMER_STATUS.SUCCESS:
+                    self.consumer.commit()
+                elif status == CONSUMER_STATUS.FAILED_STOP:
+                    # TODO: check if can be handled internally, currently message will get ack by message bus api
+                    break
+                elif status == CONSUMER_STATUS.SUCCESS_STOP:
+                    self.consumer.commit()
+                    break
+                else:
+                    retry = True
+                    continue
+                retry = False
+            except Exception as e:
+                Log.error(f"Supressing exception from message bus {e}")
+                retry = False
+
+    def start(self):
+        self.consumer = MessageBusComm(comm_type=self.comm_type, group_id=self.group_id,
```

Comment:
From line 157, Same class instantiation returns producer and here, it returns consumer. How is that possible?

Reply:
Depending on comm_type it returns producer OR consumer. On line 157 we are passing comm_type as PRODUCER & here comm_type='CONSUMER'